src/core/net/ip6.cpp

```diff
@@ -594,17 +592,34 @@ otError Ip6::ProcessReceiveCallback(const Message &aMessage, const MessageInfo &
             break;
 
         case kProtoUdp:
-            if (aMessageInfo.GetSockAddr().IsLinkLocal() ||
-                aMessageInfo.GetSockAddr().IsLinkLocalMulticast())
-            {
-                UdpHeader udp;
-                aMessage.Read(aMessage.GetOffset(), sizeof(udp), &udp);
+        {
+            UdpHeader udp;
+            aMessage.Read(aMessage.GetOffset(), sizeof(udp), &udp);
 
-                // do not pass MLE messages
-                VerifyOrExit(udp.GetDestinationPort() != Mle::kUdpPort, error = OT_ERROR_NO_ROUTE);
+            // do not pass MLE messages
+            if (udp.GetDestinationPort() == Mle::kUdpPort)
+            {
+                if (aMessageInfo.GetSockAddr().IsLinkLocal() ||
+                    aMessageInfo.GetSockAddr().IsLinkLocalMulticast())
+                {
+                    ExitNow(error = OT_ERROR_NO_ROUTE);
+                }
+            }
+            // do not pass TMF messages
+            else if (udp.GetDestinationPort() == kCoapUdpPort)
+            {
+                if (((GetInstance().mThreadNetif.GetMle().IsMeshLocalAddress(aMessageInfo.GetSockAddr()) ||
```

Comment:
This duplicates the code from [src/core/thread/thread_netif.cpp](https://github.com/openthread/openthread/blob/master/src/core/thread/thread_netif.cpp#L183).  Can we remove this duplication?